### PR TITLE
Toys API

### PR DESCRIPTION
**Pull request by PintTheDragon**

### CODE REVIEW

File: Exiled.API/Features/Toys/Light.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        public Color Color
+        {
+            get => Base.NetworkLightColor;
+            set => Base.NetworkLightColor = value;
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the light should have shadows.
+        /// </summary>
+        public bool NetworkLightShadows
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Light.cs:67]: You can rename `NetworkLightShadows` to `LightShadows` or `Shadows`

Comment (PintTheDragon) [reply Exiled.API/Features/Toys/Light.cs:76]: Oops, left that one in,

File: Exiled.API/Features/Toys/Light.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        /// <param name="toy">The toy to be wrapped.</param>
+        public Light(LightSourceToy toy)
+        {
+            Base = toy;
+        }
+
+        /// <summary>
+        /// Gets the base <see cref="LightSourceToy"/>.
+        /// </summary>
+        public LightSourceToy Base { get; } = null;
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Light.cs:39]: `= null` is redundant

File: Exiled.API/Features/Toys/Light.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        /// <summary>
+        /// Gets or sets the position of the light.
+        /// </summary>
+        public Vector3 Position { get; set; } = Vector3.zero;
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Light.cs:44]: Setter of this property won't actually change anything, please fix it

File: Exiled.API/Features/Toys/Primitive.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        public Primitive(PrimitiveObjectToy toy)
+        {
+            Base = toy;
+            Collidable = toy.NetworkPosition.x < 0 && toy.NetworkPosition.y < 0 && toy.NetworkPosition.z < 0;
+        }
+
+        /// <summary>
+        /// Gets the base <see cref="PrimitiveObjectToy"/>.
+        /// </summary>
+        public PrimitiveObjectToy Base { get; } = null;
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Primitive.cs:41]: `= null` is redundant

File: Exiled.API/Features/Toys/Primitive.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        public Vector3 Scale
+        {
+            get => Base.NetworkScale;
+            set => Base.NetworkScale = value;
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether gets or sets if the primitive can be collided with.
+        /// </summary>
+        public bool Collidable { get; set; } = true;
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Primitive.cs:91]: Again, setting this on the run, won't change anything. You can make it so when set to `true` you change the network scale and unspawn-spawn the object.

File: Exiled.API/Features/Toys/ToysHelper.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+{
+    using AdminToys;
+    using Mirror;
+
+    /// <summary>
+    /// A helper class for interacting with toys.
+    /// </summary>
+    public class ToysHelper
+    {
+        private static PrimitiveObjectToy primitiveBaseObject = null;
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/ToysHelper.cs:18]: `= null` is redundant

File: Exiled.API/Features/Toys/ToysHelper.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        private static LightSourceToy lightBaseObject = null;
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/ToysHelper.cs:19]: `= null` is redundant

File: Exiled.API/Features/Toys/Light.cs [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        /// Gets or sets the color of the primitive.
+        /// </summary>
+        public Color Color
+        {
+            get => Base.NetworkLightColor;
+            set => Base.NetworkLightColor = value;
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the light should have shadows.
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Light.cs:74]: Technically lights itself don't have shadows, I know what you mean in this but you can try changing it to something related to objects that will be illuminated by this light which is what this property do.

File: Exiled.API/Features/Toys/ToysHelper.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace Exiled.API.Features.Toys
+{
+    using AdminToys;
+    using Mirror;
+
+    /// <summary>
+    /// A helper class for interacting with toys.
+    /// </summary>
+    public class ToysHelper
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/ToysHelper.cs:16]: Could be renamed into `ToysPrefabs` or something but this name isn't bad.

**Comment by Michal78900** [comment]
You probably should also change targeted branch from `11.1` back to `dev`

File: Exiled.API/Features/Toys/Light.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    /// A wrapper class for light source Admin Toys.
+    /// </summary>
+    public class Light
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Light"/> class.
+        /// </summary>
+        public Light()
+        {
+            Base = UnityEngine.Object.Instantiate(ToysHelper.LightBaseObject);
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Light.cs:24]: `UnityEngine.` is redundant

File: Exiled.API/Features/Toys/Primitive.cs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    /// A wrapper class for primitive Admin Toys.
+    /// </summary>
+    public class Primitive
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Primitive"/> class.
+        /// </summary>
+        public Primitive()
+        {
+            Base = UnityEngine.Object.Instantiate(ToysHelper.PrimitiveBaseObject);
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Primitive.cs:25]: `UnityEngine.` is redundant

Comment (PintTheDragon) [reply Exiled.API/Features/Toys/Primitive.cs:25]: No it isn't.

File: Exiled.API/Features/Toys/Light.cs (hunk 8 above)
Comment (PintTheDragon) [reply Exiled.API/Features/Toys/Light.cs:74]: I'm not entirely sure what this property does since I haven't tested it out. Any ideas for a good name?

**Comment by PintTheDragon** [comment]
Oh my god

File: Exiled.API/Features/Toys/ToysHelper.cs (hunk 9 above)
Comment (PintTheDragon) [reply Exiled.API/Features/Toys/ToysHelper.cs:16]: Name is fine.

File: Exiled.API/Features/Toys/Light.cs (hunk 8 above)
Comment (joker-119) [reply Exiled.API/Features/Toys/Light.cs:82]: Gets or sets a value indicating whether the light should cause shadows from other objects.

ShadowEmission

**Comment by Jesus-QC** [comment]
Im missing a MovementSmoothing property.

**Info:**
You can assign it as a byte (0-255) higher values equal less smooth movement.
Use 60 for an stable smooth movement => 60 times per second.

File: Exiled.API/Features/Toys/Primitive.cs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+            set
+            {
+                Base.NetworkScale = value;
+                RefreshCollidable();
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the amount of movement smoothening on the light.
+        /// You can assign it as a byte (0-255), where higher values mean less smooth movement.
```
Comment (Jesus-QC) [inline Exiled.API/Features/Toys/Primitive.cs:96]: Remove the byte thing, it was just for you to know, just write that higher values mean less smoother and 60 is the ideal

File: Exiled.API/Features/Toys/Light.cs [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            set
+            {
+                Base.transform.position = value;
+                Base.NetworkPosition = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the amount of movement smoothening on the light.
+        /// You can assign it as a byte (0-255), where higher values mean less smooth movement.
```
Comment (Jesus-QC) [inline Exiled.API/Features/Toys/Light.cs:56]: the same as i commented in the one below

**Comment by Jesus-QC** [comment]
Ups, forgot to request changes

**Review by Jesus-QC — APPROVED** [review]
everything seems pretty clear

**Review by Jesus-QC — APPROVED** [review]
Will test it soon

File: Exiled.API/Features/Toys/ToysHelper.cs [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+                        }
+                    }
+                }
+
+                return primitiveBaseObject;
+            }
+        }
+
+        /// <summary>
+        /// Gets the base AdminToy object to instantiate when creating a new light.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Toys/ToysHelper.cs:44]: Same as above.

File: Exiled.API/Features/Toys/ToysHelper.cs [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    /// <summary>
+    /// A helper class for interacting with toys.
+    /// </summary>
+    public class ToysHelper
+    {
+        private static PrimitiveObjectToy primitiveBaseObject;
+        private static LightSourceToy lightBaseObject;
+
+        /// <summary>
+        /// Gets the base AdminToy object to instantiate when creating a new primitive.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Toys/ToysHelper.cs:22]: Same as above.

File: Exiled.API/Features/Toys/Primitive.cs [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        /// Gets or sets the amount of movement smoothening on the primitive. Higher values mean less smooth movement, and 60 is an ideal value.
+        /// </summary>
+        public byte Smoothing
+        {
+            get => Base.NetworkMovementSmoothing;
+            set => Base.NetworkMovementSmoothing = value;
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether gets or sets if the primitive can be collided with.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Toys/Primitive.cs:104]: I'd change it to _Gets or sets a value indicating whether the primitive can be collided with._

File: Exiled.API/Features/Toys/Primitive.cs [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+            get => Base.NetworkScale;
+            set
+            {
+                Base.NetworkScale = value;
+                RefreshCollidable();
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the amount of movement smoothening on the primitive. Higher values mean less smooth movement, and 60 is an ideal value.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Toys/Primitive.cs:95]: Use `<para>Higher values mean less smooth movement, and 60 is an ideal value.</para>`

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]
Other things I didn't mention were about the use of lambdas, ternary operators and explicit types.

File: Exiled.API/Features/Toys/Light.cs [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+        /// <summary>
+        /// Gets or sets the position of the light.
+        /// </summary>
+        public Vector3 Position
+        {
+            get => Base.NetworkPosition;
+            set
+            {
+                Base.transform.position = value;
+                Base.NetworkPosition = value;
```
Comment (Jesus-QC) [inline Exiled.API/Features/Toys/Light.cs:50]: not needed

File: Exiled.API/Features/Toys/Primitive.cs [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        /// <summary>
+        /// Gets or sets the scale of the primitive.
+        /// </summary>
+        public Vector3 Scale
+        {
+            get => Base.NetworkScale;
+            set
+            {
```
Comment (Jesus-QC) [inline Exiled.API/Features/Toys/Primitive.cs:88]: Base.Transform.localScale too (NetworkScale is the syncvar)

**Review by Jesus-QC — CHANGES_REQUESTED** [review]

File: Exiled.API/Features/Toys/Light.cs (hunk 18 above)
Comment (PintTheDragon) [reply Exiled.API/Features/Toys/Light.cs:50]: Why not?

Comment (Jesus-QC) [reply Exiled.API/Features/Toys/Light.cs:50]: `LigthSourceToy` is derivated from `AdminToyBase` which has a method called `UpdatePositionServer` called on **LateUpdate**

Comment (Jesus-QC) [inline Exiled.API/Features/Toys/Light.cs:50]: Dont use network stuff, use transform, the network stuff is automatically synced, therefore if you change it directly via network... it wont change its transform so youre desyncing it from the server some frames until the updateserver method is called and its network things are synced with its transform

**Review by Jesus-QC — CHANGES_REQUESTED** [review]
Position, Rotation and Scale shouldn't be directly changed with Network... however the rest can be safely changed with that.

File: Exiled.API/Features/Toys/Light.cs [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+        /// Gets the base <see cref="AdminToys.LightSourceToy"/>.
+        /// </summary>
+        public LightSourceToy Base { get; }
+
+        /// <summary>
+        /// Gets or sets the position of the light.
+        /// </summary>
+        public Vector3 Position
+        {
+            get => Base.NetworkPosition;
```
Comment (Jesus-QC) [inline Exiled.API/Features/Toys/Light.cs:46]: NetworkPosition will return the synced position of the object, not the real one always.

**Review by Jesus-QC — CHANGES_REQUESTED** [review]

**Comment by Jesus-QC** [comment]
Why tf it only shows one req change.

Pint, change all NetworkPosition, NetworkRotation and NetworkScale to it transform one which is the real one.

*`AdminToyBase`*
image
image

File: Exiled.API/Features/Toys/Primitive.cs [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+
+        /// <summary>
+        /// Gets or sets the scale of the primitive.
+        /// </summary>
+        public Vector3 Scale
+        {
+            get => Base.transform.localScale;
+            set
+            {
+                Base.transform.localScale = value;
```
Comment (Jesus-QC) [inline Exiled.API/Features/Toys/Primitive.cs:91]: Unnecessary

**Review by Jesus-QC — APPROVED** [review]
everything seems cool

File: Exiled.API/Features/Toys/Primitive.cs [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Primitive"/> class from a <see cref="AdminToys.PrimitiveObjectToy"/>.
+        /// </summary>
+        /// <param name="toy">The <see cref="AdminToys.PrimitiveObjectToy"/> to be wrapped.</param>
+        public Primitive(PrimitiveObjectToy toy)
+        {
+            Base = toy;
+
+            var position = toy.transform.position;
+            Collidable = position.x < 0 && position.y < 0 && position.z < 0;
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Primitive.cs:39]: position?? Didn't you mean scale?

**Review by joker-119 — CHANGES_REQUESTED** [review]
Why not have a generic Toy class (abstract) that acts as a wrapper for AdminToyBase
Then have both Primitive and Light inherit that Toy class?
The Toy class itself can handle the Position, Collidable, Scale, Smoothing, Spawn and Unspawn
Then the Light and Primitive classes only need methods/properties for themselves specifically, and plugins (like MER and others) can just store a list of all toys they have spawned in a List<Toy> rather than needing a List<Light> and List<Primitive>

Similar to how the Item API works.

**Comment by Jesus-QC** [comment]
> Why not have a generic Toy class (abstract) that acts as a wrapper for AdminToyBase Then have both Primitive and Light inherit that Toy class? The Toy class itself can handle the Position, Collidable, Scale, Smoothing, Spawn and Unspawn Then the Light and Primitive classes only need methods/properties for themselves specifically, and plugins (like MER and others) can just store a list of all toys they have spawned in a List rather than needing a List and List
> 
> Similar to how the Item API works.

Yeah, that would be awesome

**Comment by Jesus-QC** [comment]
Im helping with it. I have recoded everything using an abstract class as a wrapper for admin toy base, @PintTheDragon hope you dont mind :).

**Comment by Jesus-QC** [comment]
Spawn a light:
```cs
var l = Light.Create();
l.Spawn();
```

Spawn a primitive:
```cs
var p = Primitive.Create();
p.Spawn();
```

**Review by NaoUnderscore — APPROVED** [review]
Seems fine

File: Exiled.API/Features/Toys/Primitive.cs [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+        /// <summary>
+        /// Gets or sets the scale of the primitive.
+        /// </summary>
+        public override Vector3 Scale
+        {
+            get => AdminToyBase.transform.localScale;
+            set
+            {
+                AdminToyBase.transform.localScale = value;
+                RefreshCollidable();
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/Primitive.cs:80]: I think this won't do shit, you need to actually UnSpawn and Spawn an object but I'm not sure

**Comment by Michal78900** [comment]
I would like to have a list of all spawned primitves, similar to what `ShootingTarget` and other stuff have. Actually the `ShootingTarget` is now a toy, so it also should be a part of this API

Since the abstract `AdminToy` class exists, you could merge `ToysHelper` into it, but it isn't a big issue.

File: Exiled.API/Features/Toys/AdminToy.cs [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+        public AdminToy(AdminToyBase toyAdminToyBase, AdminToyType type)
+        {
+            AdminToyBase = toyAdminToyBase;
+            ToyType = type;
+        }
+
+        /// <summary>
+        /// Gets the original <see cref="AdminToys.AdminToyBase"/>.
+        /// </summary>
+        public AdminToyBase AdminToyBase { get; }
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/AdminToy.cs:34]: Everything in EXILED uses a `Base` word, I would use it here too

Comment (Michal78900) [reply Exiled.API/Features/Toys/AdminToy.cs:34]: No wait, not this but in `Light` class

Comment (Jesus-QC) [reply Exiled.API/Features/Toys/AdminToy.cs:34]: oh wait, this is the admintoy class, since its derivated from it we cant use the same name

Comment (Jesus-QC) [reply Exiled.API/Features/Toys/AdminToy.cs:34]: So if you have a light `li` you can get `li.Base` and `li.AdminToyBase`

**Comment by Jesus-QC** [comment]
ups

**Comment by Jesus-QC** [comment]
To do:
```yaml
- Map.Toys (working on it rn)
- Move ShootingTarget from features to here and make it derivated from admintoy
```

**Comment by Jesus-QC** [comment]
if anyone wants to do it, go ahead

**Review by NaoUnderscore — APPROVED** [review]
Looks good

File: Exiled.API/Features/Toys/AdminToy.cs [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+        /// </para>
+        /// </summary>
+        public byte MovementSmoothing
+        {
+            get => AdminToyBase.MovementSmoothing;
+            set => AdminToyBase.NetworkMovementSmoothing = value;
+        }
+
+        /// <summary>
+        /// Spawns the light into the game. Use <see cref="UnSpawn"/> to remove it.
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/AdminToy.cs:82]: It should be `toy` instead `light`.

File: Exiled.API/Features/Toys/AdminToy.cs [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+        /// </summary>
+        public void Spawn()
+        {
+            NetworkServer.Spawn(AdminToyBase.gameObject);
+            Map.ToysValue.Add(this);
+        }
+
+        /// <summary>
+        /// Removes the light from the game. Use <see cref="Spawn"/> to bring it back.
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/AdminToy.cs:91]: It should be `toy` instead `light`.

File: Exiled.API/Features/Toys/ToysHelper.cs [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+namespace Exiled.API.Features.Toys
+{
+    using AdminToys;
+    using Mirror;
+    using UnityEngine;
+
+    /// <summary>
+    /// A helper class for interacting with toys.
+    /// </summary>
+    public class ToysHelper
```
Comment (Michal78900) [inline Exiled.API/Features/Toys/ToysHelper.cs:17]: This class should probably be `static`

**Comment by Michal78900** [comment]
I don't understand why you removed `Get` method from ShootingTarget it was very useful and guaraneteed that no extra instances of the object will be created. I would add it back and add it to every toy.

**Comment by Michal78900** [comment]
Taka a look here: https://github.com/Exiled-Team/EXILED/pull/1038
Constructors are internal, and there is a `Get` method. In our case we would have both `Get` and `Create` methods

I will try making a PR into this, I think it would make more sense then

**Review by Michal78900 — APPROVED** [review]

**Review by NaoUnderscore — APPROVED** [review]

**Review by Jesus-QC — APPROVED** [review]

**Pull request merged**